feat: add validation function for authorizationArgs

JohnStarich opened the pull request:
Adds validation to authorization args. Effectively moves notarize-specific validation out of electron-packager, but enables others to call `validateAuthorizationArgs` in their own package.

This PR is part of enabling API key auth in electron-packager: https://github.com/electron/electron-packager/pull/1127

This is the first time I'm using TypeScript, so if anything looks off, that's probably why! :smile:

malept commented on src/index.ts:
```diff
@@ -36,8 +36,59 @@ export type NotarizeWaitOptions = NotarizeResult & NotarizeCredentials;
 
-function authorizationArgs(opts: NotarizeCredentials): string[] {
-  if ('appleId' in opts) {
+function isPasswordCredentials(opts: NotarizeCredentials): opts is NotarizePasswordCredentials {
```
> `(...): opts is NotarizePasswordCredentials`
I have never seen that syntax before, can you link me to the handbook reference?

JohnStarich replied:
Sure thing: https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards
From what I can tell, it provides a type hint to TypeScript for if statements using the condition. Seems it's typically used with [Union types](https://www.typescriptlang.org/docs/handbook/advanced-types.html#union-types) in the examples.

malept commented on src/index.ts:
```diff
@@ -36,8 +36,59 @@ export type NotarizeWaitOptions = NotarizeResult & NotarizeCredentials;
+  const isPassword = isPasswordCredentials(opts);
+  const isApiKey = isApiKeyCredentials(opts);
+  if (isPassword && isApiKey) {
+    throw new Error('Mixed authentication properties appleId and appleApiKey');
```
```suggestion
    throw new Error('Cannot use both password credentials and API key credentials at once');
```

malept commented on src/index.ts:
```diff
@@ -36,8 +36,59 @@ export type NotarizeWaitOptions = NotarizeResult & NotarizeCredentials;
+      );
+    }
+
+    if (!passwordCreds.appleIdPassword) {
```
```suggestion
    else if (!passwordCreds.appleIdPassword) {
```

With appropriate formatting, of course.

JohnStarich replied:
Happy to add the `else` bits, but I think it's worth pointing out the `throw`s make an `else` ineffectual here.

malept replied:
Yep, I get it's a style thing, but that's what I generally prefer to see as it makes more logical sense to me.

malept commented on src/index.ts:
```diff
@@ -36,8 +36,59 @@ export type NotarizeWaitOptions = NotarizeResult & NotarizeCredentials;
+      );
+    }
+
+    if (!apiKeyCreds.appleApiIssuer) {
```
```suggestion
    else if (!apiKeyCreds.appleApiIssuer) {
```

With appropriate formatting, of course.

malept commented on src/index.ts:
```diff
@@ -36,8 +36,59 @@ export type NotarizeWaitOptions = NotarizeResult & NotarizeCredentials;
+}
+
+function isApiKeyCredentials(opts: NotarizeCredentials): opts is NotarizeApiKeyCredentials {
+  const creds = opts as NotarizeApiKeyCredentials;
```
```suggestion
function isApiKeyCredentials(creds: NotarizeApiKeyCredentials): boolean {;
```

JohnStarich replied:
I can change this too. What are your thoughts on keeping the type guards? From what I read, the type predicate (`x is Type` syntax) in the handbook is recommended for this use case since `NotarizeApiKeyCredentials` is a union type:
```ts
export type NotarizeCredentials = NotarizePasswordCredentials | NotarizeApiKeyCredentials;
```

malept replied:
OK, now that I've read the handbook reference, I agree that we should keep the type guards. I'd still rename the parameter.

malept commented on src/index.ts:
```diff
@@ -36,8 +36,59 @@ export type NotarizeWaitOptions = NotarizeResult & NotarizeCredentials;
+  const apiKeyCreds = opts as NotarizeApiKeyCredentials;
+
+  const isPassword = isPasswordCredentials(opts);
+  const isApiKey = isApiKeyCredentials(opts);
```
```suggestion
  const isApiKey = isApiKeyCredentials(apiKeyCreds);
```

JohnStarich replied:
Oh oops, I intended to place the type-casted `const`s in their if's instead. So in this case the `opts` parameter is the union type and `isApiKey` will provide the type hint we can use to access its properties.

malept commented on src/index.ts:
```diff
@@ -36,8 +36,59 @@ export type NotarizeWaitOptions = NotarizeResult & NotarizeCredentials;
+    }
+    return passwordCreds;
+  }
+  if (isApiKey) {
```
```suggestion
  else if (isApiKey) {
```

JohnStarich commented:
Updated & responded! Let me know when the test framework is ready and I'll push up some tests too 👍 

I also attempted to swap out the type guards `opts is NotarizePasswordCredentials` with `boolean` and ran a build. It seems unhappy without the type hints; here's the errors it generated:

<details>
<summary>

```ts
src/index.ts:87:35 - error TS2339: Property 'appleId' does not exist on type 'NotarizeCredentials'.
  Property 'appleId' does not exist on type 'NotarizeApiKeyCredentials'.

87     return ['-u', makeSecret(opts.appleId), '-p', makeSecret(opts.appleIdPassword)];
                                     ~~~~~~~
```
</summary>

```
➜ electron-notarize npm run build

> electron-notarize@0.0.0-development build /Users/johnstarich/Documents/Projects/electron-notarize
> tsc

src/index.ts:87:35 - error TS2339: Property 'appleId' does not exist on type 'NotarizeCredentials'.
  Property 'appleId' does not exist on type 'NotarizeApiKeyCredentials'.

87     return ['-u', makeSecret(opts.appleId), '-p', makeSecret(opts.appleIdPassword)];
                                     ~~~~~~~

src/index.ts:87:67 - error TS2339: Property 'appleIdPassword' does not exist on type 'NotarizeCredentials'.
  Property 'appleIdPassword' does not exist on type 'NotarizeApiKeyCredentials'.

87     return ['-u', makeSecret(opts.appleId), '-p', makeSecret(opts.appleIdPassword)];
                                                                     ~~~~~~~~~~~~~~~

src/index.ts:91:23 - error TS2339: Property 'appleApiKey' does not exist on type 'NotarizeCredentials'.
  Property 'appleApiKey' does not exist on type 'NotarizePasswordCredentials'.

91       makeSecret(opts.appleApiKey),
                         ~~~~~~~~~~~

src/index.ts:93:23 - error TS2339: Property 'appleApiIssuer' does not exist on type 'NotarizeCredentials'.
  Property 'appleApiIssuer' does not exist on type 'NotarizePasswordCredentials'.

93       makeSecret(opts.appleApiIssuer),
                         ~~~~~~~~~~~~~~


Found 4 errors.

npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! electron-notarize@0.0.0-development build: `tsc`
npm ERR! Exit status 2
npm ERR!
npm ERR! Failed at the electron-notarize@0.0.0-development build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/johnstarich/.npm/_logs/2020-03-04T02_20_27_697Z-debug.log
➜
```
</details>

malept commented:
I've merged #36 to master which adds basic Jest support, among other things.

malept commented:
> (will squash later)

Not necessary, I'll squash & merge when it's approved.

JohnStarich commented:
Thanks for the quick turn-around on the test framework and reviews! (And your patience with me :wink:) I've pushed some new tests and addressed the remaining comments.

On the tests, I've covered each branch in `validateAuthorizationArgs`, though I stopped short of testing `authorizationArgs` since I'm not sure how to test unexported funcs.

malept commented:
>  I'm not sure how to test unexported funcs.

Personally, I'd add an `src/validate-args.ts` and move all of the functions you added to `src/index.ts`, in order to add tests for them as necessary. Then just re-export `validateAuthorizationArgs`.

JohnStarich commented:
Works for me! Great suggestion, I'd never heard of "re-exporting" before -- very clean solution.

JohnStarich commented:
All set on this PR. Please let me know if there's anything I've missed!
